flocker/ca/test/test_ca.py

```diff
@@ -16,13 +16,296 @@
+            b"Permission denied {path}"
+        ).format(path=crt_path.path)
+        self.assertEqual(str(e), expected)
+
+    @not_root
+    @skip_on_broken_permissions
+    def test_load_error_on_unreadable_key_file(self):
+        """
+        A ``PathError`` is raised if the key file path given to
+        ``UserCredential.from_path`` cannot be opened for reading.
+        """
+        path = FilePath(self.mktemp())
+        path.makedirs()
+        cert_file = b"{user}.crt".format(user=self.username)
+        key_file = b"{user}.key".format(user=self.username)
+        crt_path = path.child(cert_file)
+        crt_file = crt_path.open(b'w')
+        crt_file.write(b"dummy")
+        crt_file.close()
+        key_path = path.child(key_file)
+        key_file = key_path.open(b'w')
+        key_file.write(b"dummy")
+        key_file.close()
+        # make file unreadable
+        key_path.chmod(64)
+        e = self.assertRaises(
+            PathError, UserCredential.from_path, path, self.username
+        )
+        expected = (
+            b"Private key file could not be opened. "
+            b"Permission denied {path}"
+        ).format(path=key_path.path)
+        self.assertEqual(str(e), expected)
+
+    def test_certificate_subject_username(self):
+        """
+        A cert written by ``UserCredential.initialize`` has the
+        subject common name "user-{user}" where {user} is the username
+        supplied during the certificate's creation.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        cert = uc.credential.certificate.original
+        subject = cert.get_subject()
+        self.assertEqual(subject.CN, b"user-{user}".format(user=uc.username))
+
+    def test_certificate_ou_matches_ca(self):
+        """
+        A cert written by ``UserCredential.initialize`` has the issuing
+        authority's common name as its organizational unit name.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        cert = uc.credential.certificate.original
+        issuer = cert.get_issuer()
+        subject = cert.get_subject()
+        self.assertEqual(
+            issuer.CN,
+            subject.OU
+        )
+
+    def test_certificate_is_signed_by_ca(self):
+        """
+        A cert written by ``UserCredential.initialize`` is validated
+        as being signed by the certificate authority.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        cert = uc.credential.certificate.original
+        issuer = cert.get_issuer()
+        self.assertEqual(
+            issuer.CN,
+            self.ca.credential.certificate.getSubject().CN
+        )
+
+    def test_certificate_expiration(self):
+        """
+        A cert written by ``UserCredential.initialize`` has an expiry
+        date 20 years from the date of signing.
+        """
+        today = datetime.datetime.now()
```

Comment:
I think you said there's a branch that fixes the time math in these tests.  Can you link to it from here?